<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>An Introduction to FastqCleaner • FastqCleaner</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="An Introduction to FastqCleaner">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">FastqCleaner</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.5.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Overview.html">An Introduction to FastqCleaner</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>An Introduction to FastqCleaner</h1>
                        <h4 class="author">Leandro Roser</h4>
            <address class="author_afil">
      IIB-INTECH - UNSAM<br><h4 class="author">Fernán Agüero</h4>
            <address class="author_afil">
      IIB-INTECH - UNSAM<br><h4 class="author">Daniel Sánchez</h4>
            <address class="author_afil">
      IIB-INTECH - UNSAM<br><a class="author_email" href="mailto:#"></a><a href="mailto:learoser@gmail.com" class="email">learoser@gmail.com</a>
      </address>
                  
            <h4 class="date">Feb 15, 2020</h4>
      
      
      <div class="hidden name"><code>Overview.Rmd</code></div>

    </address>
</address>
</div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p><em><a href="https://bioconductor.org/packages/3.9/FastqCleaner">FastqCleaner</a></em> is an interactive application for both quality control and pre-processing of FASTQ files. The interface, based on a Bioconductor back-end, provides diagnostic information for the input and output data and a series of filtering and trimming operations in an interactive framework.</p>
    </div>
    
<div id="launching-the-application" class="section level1">
<h1 class="hasAnchor">
<a href="#launching-the-application" class="anchor"></a>Launching the application</h1>
<p>The interactive application can be launched in R with the following command:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">'FastqCleaner'</span>)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="kw"><a href="../reference/launch_fqc.html">launch_fqc</a></span>()</span></code></pre></div>
<p>As an alternative method, an RStudio addin (RStudio version 0.99.878 or higher required) installed with the package can be found in the Addins menu (Figure 1). This button allows the direct launch of the application with a single click.</p>
<p><img src="images/Fig1.png"><strong>Figure 1</strong>: addin of the app in RStudio (RStudio version &gt;= 0.99.878 required)</p>
<p> </p>
</div>
<div id="description-of-the-application" class="section level1">
<h1 class="hasAnchor">
<a href="#description-of-the-application" class="anchor"></a>Description of the application</h1>
<p>The application contains three main panels, as described below.</p>
<div id="first-panel" class="section level2">
<h2 class="hasAnchor">
<a href="#first-panel" class="anchor"></a>First panel</h2>
<p>The first panel includes two elements: a dashboard for selection of trimming and filtering operations, and a menu for selection of the input file/s (Fig. 2).</p>
<p><img src="images/Fig2.png"></p>
<p><strong>Figure 2</strong>: Panel 1, with each compontent indicated with a number. See explanation for each element below</p>
<p> </p>
<div id="selecting-operations" class="section level3">
<h3 class="hasAnchor">
<a href="#selecting-operations" class="anchor"></a>Selecting operations</h3>
<p>The “operations menu” (Fig. 2, elements 1 to 8) shows the available operations for file processing:</p>
<ol style="list-style-type: decimal">
<li><p><strong>Remove by N(s)</strong>: removes sequences with a number of Ns (non identified bases) above a selected threshold value</p></li>
<li><p><strong>Remove low complexity sequences</strong>: remove sequences with a value of complexity above a threshold value</p></li>
<li><p><strong>Remove adapters</strong>: removes adapters and partial adapters. Adapter sequences from both ends of single or paired read reads can be selected. Sequences can be reverse-complemented before processing. The program also allows to consider indels and/or anchored adapters.</p></li>
<li><p><strong>Filter by average quality</strong>: computes the average quality of sequences and removes those with a value below a given threshold</p></li>
<li><p><strong>Trim low quality 3’ tails</strong>: removes the 3’ tails of sequences that are below a given threshold</p></li>
<li><p><strong>Trim 3’ or 5’ by a fixed number</strong>: removes a fixed number of bases from the 3’ and/or 5’ ends in the complete set of sequences</p></li>
<li><p><strong>Filter sequences by length</strong>: removes all the sequences with a number of bases below a threshold value</p></li>
<li><p><strong>Remove duplicated sequences</strong>: removes duplicated reads, conserving only one copy of each sequence present in the file</p></li>
</ol>
</div>
<div id="loading-files" class="section level3">
<h3 class="hasAnchor">
<a href="#loading-files" class="anchor"></a>Loading files</h3>
<p>The “file selection menu” (Fig. 2, elements 9 to 17) contains options to handle the input file (type of file, file selection), buttons to run, clear and reset the aplication, and the “advanced” submenu:</p>
<ol start="9" style="list-style-type: decimal">
<li><p><strong>Single-end reads / paired-end reads</strong>: type of input files</p></li>
<li><p><strong>“FILE” button</strong>: to select an input file</p></li>
<li><p><strong>“RUN!” button</strong>: to run the program</p></li>
<li><p><strong>Output format</strong>: to select whether the output file should be compressed (.gz) or not</p></li>
<li><p><strong>“CLEAR” button</strong>: to clear the configuration of the operations menu that have been selected in the first panel, but keeping the input file(s)</p></li>
<li><p><strong>“RESET” button</strong>: to restart the application, removing the input file(s) and the selected configurations</p></li>
<li><p><strong>Selection notificator</strong>: information of the path of the selected file/s</p></li>
<li><p><strong>Encoding notificator</strong>: information of the input file/s encoding</p></li>
<li><p><strong>Advanced options button</strong>: to select a custom encoding and set the number of reads included in each chunk for processing, as described below</p></li>
</ol>
</div>
<div id="advanced-options" class="section level3">
<h3 class="hasAnchor">
<a href="#advanced-options" class="anchor"></a>Advanced options</h3>
<p>The “advanced options submenu” (Fig. 3) allows to customize some fine aspects of the trimming and filtering process:</p>
<p><img src="images/Fig3.png"></p>
<p><strong>Figure 3</strong>: Advanced options submenu</p>
<p> </p>
<ol start="18" style="list-style-type: decimal">
<li><p><strong>Encoding menu</strong>: in addition to the default approach used by the program (auto-detection of file encoding), users can select a standard encoding from a list</p></li>
<li><p><strong>Chunk size</strong>: the program takes this number of reads at random from the file (default: 1000000), for encoding detection</p></li>
</ol>
<p> </p>
</div>
</div>
<div id="second-panel" class="section level2">
<h2 class="hasAnchor">
<a href="#second-panel" class="anchor"></a>Second panel</h2>
<p>The second panel (“file operations” panel, Fig. 4) shows the operations that were sucessfuly performed on the input file after running the program.</p>
<p><img src="images/Fig4.png"></p>
<p><strong>Figure 4</strong>: File operations panel, with its elements</p>
<p> </p>
<p>The panel contains the following elements:</p>
<ol start="20" style="list-style-type: decimal">
<li><p><strong>Files location</strong>: location of input and output files</p></li>
<li><p><strong>Operations performed</strong>: operations perfomed on the input file. Each individual display indicates the number of reads that passed the corresponding filter</p></li>
</ol>
<p> </p>
</div>
<div id="third-panel" class="section level2">
<h2 class="hasAnchor">
<a href="#third-panel" class="anchor"></a>Third panel</h2>
<p>The third panel ( “live results” panel, Fig. 5) shows interactive diagnostics plots for both input and output files. The program takes a random sample of reads for construction of the plots (default: 10000 reads).</p>
<p><img src="images/Fig5.png"></p>
<p><strong>Figure 5</strong>: Live results panel</p>
<p> </p>
<p>The panel includes the following options in the menu located on the left:</p>
<ol start="22" style="list-style-type: decimal">
<li><p><strong>Sample size</strong>: the sample size used for construction of the plots. Default: 10000 reads</p></li>
<li><p><strong>Input / output</strong>: show diagnostics plots for input or output files?</p></li>
<li>
<p><strong>Diagnostics plots</strong>: the plot to be shown, that can be one of the following:</p>
<ul>
<li><p><em>Per cycle quality</em>: quality plots across reads for each cycle (i.e., sequence position)</p></li>
<li><p><em>Per cycle mean quality</em>: average quality across reads per base, for each cycle (i.e., sequence position)</p></li>
<li><p><em>Mean quality distribution</em>: Quality distribution, using for the construction of the histogram the mean quality of each read</p></li>
<li><p><em>% reads with Phred scores &gt; threshold</em>: % of reads with all the quality values &gt; threshold</p></li>
<li><p><em>Per cycle base proportion</em>: Proportion of each base (average across reads) in each cycle. It also shows the proporion of N’s</p></li>
<li><p><em>CG content</em>: % CG and % AT (average across reads) for each cycle</p></li>
<li><p><em>CG content distribution over all reads</em>: histogram for % reads with a given % CG</p></li>
<li><p><em>Read length distribution</em>: % reads vs read length (bp)</p></li>
<li><p><em>Read ocurrence distribution</em>: % reads that ocurr at different frequencies values in the file. The plot also includes a table</p></li>
<li><p><em>Relative k-mer diversity</em>: unique k-mers / all posible kmers for each cycle</p></li>
</ul>
</li>
<li><p><strong>Select k-mer size</strong>: k-mer size for the k-mers frequency plot</p></li>
<li><p><strong>Top sequences in duplication level analysis</strong>: a list of duplicated sequences, ordered from high to low duplication level, can be desplegated from the “read ocurrence distribution” plot. The number selected here indicates how many sequences should be shown. Note that the frequency of reads are relative to the sample size selected (i.e., fold-times in relation to those reads present only once in the sample)</p></li>
<li><p><strong>Plot panel</strong></p></li>
</ol>
<p>  </p>
</div>
</div>
<div id="a-worked-example-fastq-processing-in-a-nutshell" class="section level1">
<h1 class="hasAnchor">
<a href="#a-worked-example-fastq-processing-in-a-nutshell" class="anchor"></a>A worked example: FASTQ processing in a nutshell</h1>
<p>A sample FASTQ (gz-compressed) file ‘example.fastq.gz’ can be downloaded with the following command in R:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="kw"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span>(<span class="st">"https://goo.gl/hb4Kr9"</span>, <span class="st">"example_fastq.gz"</span>)</span></code></pre></div>
<p>A direct download is provided in <a href="https://goo.gl/hb4Kr9"> this link </a>.  </p>
<p>A tipical <em><a href="https://bioconductor.org/packages/3.9/FastqCleaner">FastqCleaner</a></em> workflow starts with the input file/s upload (Fig. 6).</p>
<p><img src="images/Fig6.png"></p>
<p><strong>Figure 6</strong>: File input menu. The example shows a single-end reads case (sample file ‘example.fastq.gz’). For paired-end reads, the selection of the corresponding library type generates an additional button to upload the second file.</p>
<p> </p>
<p>The file encoding is automatically detected by the program, but it can also be manually specified in the advanced submenu (Fig. 7). This menu also offers an option to customize the chunk size used for processing.</p>
<p><img src="images/Fig7.png"></p>
<p><strong>Figure 7</strong>: Advanced submanu</p>
<p> </p>
<p>Next, the operations to be performed on the input file are selected from the operations menu (Fig. 8).</p>
<p><img src="images/Fig8.png"></p>
<p><strong>Figure 8</strong>: Selection of operations. A dialog box shows the input expected for the program. To use a filter, the “Use filter?” checkbox must be checked. A filter in use is indicated with a checkmark in the filter box</p>
<p> </p>
<p>The program then starts to run after pressing the “RUN!” button (Fig. 9).</p>
<p><img src="images/Fig9.png"></p>
<p><strong>Figure 9</strong>: “RUN!” button action</p>
<p> </p>
<p>Post-processing results are shown in the second panel (Fig. 10).</p>
<p><img src="images/Fig10.png"></p>
<p><strong>Figure 10</strong>: Second panel of the app, showing the operations performed and the paths of the input and output files</p>
<p> </p>
<p>The type of plot to be displayed and the options for the construction of the plot are available in the third panel (Fig. 11). This panel also show the selected plot/s.</p>
<p><img src="images/Fig11.png"></p>
<p><strong>Figure 11</strong>: Third panel, showing as example a “CG” content plot. for the output file</p>
<p> </p>
<p>To clean the operations, for example to run a different configuration, the “CLEAN” i(Fig. 11) must be pressed. The “RESET” button (Fig. 11) restarts the interface.</p>
<p> </p>
<p>Additional help can be found in the “help” button located at the top-right of the app (Fig. 12).</p>
<p><img src="images/Fig12.png"></p>
<p><strong>Figure 12</strong>: help button. A webpage with information will be open</p>
<p> </p>
</div>
<div id="advanced-use-of-the-package" class="section level1">
<h1 class="hasAnchor">
<a href="#advanced-use-of-the-package" class="anchor"></a>Advanced use of the package</h1>
<p><em><a href="https://bioconductor.org/packages/3.9/FastqCleaner">FastqCleaner</a></em> separates the interface from the implementation. In consequence, the processing functions of the package can be used as standard functions from the command line. Most of the functions make intensive use of <em><a href="https://bioconductor.org/packages/3.9/Biostrings">Biostrings</a></em> and <em><a href="https://bioconductor.org/packages/3.9/ShortRead">ShortRead</a></em>. Trimming and filtering is performed on ShortReadQ objects. A complete documentation for the functions is available in <a href="docs/reference/index.html">this link</a></p>
<p>The functions included in the package are described in the following section.</p>
<p> </p>
<div id="main-functions" class="section level2">
<h2 class="hasAnchor">
<a href="#main-functions" class="anchor"></a>Main functions</h2>
<ul>
<li><strong>adapter_filter</strong></li>
</ul>
<p>Based on the <em><a href="https://bioconductor.org/packages/3.9/Biostrings">Biostrings</a></em> trimLRPatterns functions. It can remove adapters and partial adapters from the 3’ and 5’ sequence ends. Adapters can be anchored or not. When indels are allowed, the method is based on the “edit distance” of the sequences.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="co">### Examples</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="st">"Biostrings"</span>)</span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="st">"ShortRead"</span>)</span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="st">"FastqCleaner"</span>)</span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="co"># create sequences</span></span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">10</span>) </span>
<span id="cb5-3"><a href="#cb5-3"></a><span class="co"># nota that the use of set.seed before the call to the </span></span>
<span id="cb5-4"><a href="#cb5-4"></a><span class="co"># random generators allows reproducibility of the</span></span>
<span id="cb5-5"><a href="#cb5-5"></a><span class="co"># examples</span></span>
<span id="cb5-6"><a href="#cb5-6"></a></span>
<span id="cb5-7"><a href="#cb5-7"></a>input &lt;-<span class="st"> </span><span class="kw"><a href="../reference/random_seq.html">random_seq</a></span>(<span class="dv">6</span>, <span class="dv">43</span>)</span>
<span id="cb5-8"><a href="#cb5-8"></a>input</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 6
##     width seq
## [1]    43 TGGTCCGGTGTTCTGGCGGAATAGGTACAGTCCAGTAATTGCC
## [2]    43 TCCCGCAGACGCTGGGTCCGGAATGCCCTTTCTGAGCAGCTCC
## [3]    43 AGCCGTTTGACTTCGCGGAAAGTGAACTTAGATTCGGTCCTGA
## [4]    43 AACACGGTACTTCCACAGTCAACCCGCCGACTTGGAGAATTTA
## [5]    43 TTAGCCGGGCGGTTATTCCCCTAGTGATCTTACTAAGATTTGC
## [6]    43 AATACCTAAGCGAAGTGACAGATATGTTCGTCATTCATCCAGG</code></pre>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="co"># create qualities of width 50</span></span>
<span id="cb7-2"><a href="#cb7-2"></a><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">10</span>)</span>
<span id="cb7-3"><a href="#cb7-3"></a>input_q &lt;-<span class="st"> </span><span class="kw"><a href="../reference/random_qual.html">random_qual</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">30</span>,<span class="dv">40</span>), <span class="dt">slength =</span> <span class="dv">6</span>, <span class="dt">swidth =</span> <span class="dv">50</span>, </span>
<span id="cb7-4"><a href="#cb7-4"></a><span class="dt">encod =</span> <span class="st">"Sanger"</span>)</span>
<span id="cb7-5"><a href="#cb7-5"></a></span>
<span id="cb7-6"><a href="#cb7-6"></a><span class="co"># create names</span></span>
<span id="cb7-7"><a href="#cb7-7"></a>input_names &lt;-<span class="st"> </span><span class="kw"><a href="../reference/seq_names.html">seq_names</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(input))</span>
<span id="cb7-8"><a href="#cb7-8"></a></span>
<span id="cb7-9"><a href="#cb7-9"></a></span>
<span id="cb7-10"><a href="#cb7-10"></a><span class="co">### FULL ADAPTER IN 3'</span></span>
<span id="cb7-11"><a href="#cb7-11"></a>adapter &lt;-<span class="st"> "ATCGACT"</span></span>
<span id="cb7-12"><a href="#cb7-12"></a></span>
<span id="cb7-13"><a href="#cb7-13"></a><span class="co"># Create sequences with adapter</span></span>
<span id="cb7-14"><a href="#cb7-14"></a>my_seqs &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(input, adapter)</span>
<span id="cb7-15"><a href="#cb7-15"></a>my_seqs &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html">DNAStringSet</a></span>(my_seqs)</span>
<span id="cb7-16"><a href="#cb7-16"></a>my_seqs</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 6
##     width seq
## [1]    50 TGGTCCGGTGTTCTGGCGGAATAGGTACAGTCCAGTAATTGCCATCGACT
## [2]    50 TCCCGCAGACGCTGGGTCCGGAATGCCCTTTCTGAGCAGCTCCATCGACT
## [3]    50 AGCCGTTTGACTTCGCGGAAAGTGAACTTAGATTCGGTCCTGAATCGACT
## [4]    50 AACACGGTACTTCCACAGTCAACCCGCCGACTTGGAGAATTTAATCGACT
## [5]    50 TTAGCCGGGCGGTTATTCCCCTAGTGATCTTACTAAGATTTGCATCGACT
## [6]    50 AATACCTAAGCGAAGTGACAGATATGTTCGTCATTCATCCAGGATCGACT</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="co"># create ShortReadQ object</span></span>
<span id="cb9-2"><a href="#cb9-2"></a>my_read &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ShortRead/man/ShortReadQ-class.html">ShortReadQ</a></span>(<span class="dt">sread =</span> my_seqs, <span class="dt">quality =</span> input_q, <span class="dt">id =</span> input_names)</span>
<span id="cb9-3"><a href="#cb9-3"></a></span>
<span id="cb9-4"><a href="#cb9-4"></a><span class="co"># trim adapter</span></span>
<span id="cb9-5"><a href="#cb9-5"></a>filtered &lt;-<span class="st"> </span><span class="kw"><a href="../reference/adapter_filter.html">adapter_filter</a></span>(my_read, <span class="dt">Lpattern =</span> adapter)</span>
<span id="cb9-6"><a href="#cb9-6"></a><span class="kw"><a href="https://rdrr.io/pkg/ShortRead/man/ShortRead-class.html">sread</a></span>(filtered)</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 6
##     width seq
## [1]    50 TGGTCCGGTGTTCTGGCGGAATAGGTACAGTCCAGTAATTGCCATCGACT
## [2]    50 TCCCGCAGACGCTGGGTCCGGAATGCCCTTTCTGAGCAGCTCCATCGACT
## [3]    50 AGCCGTTTGACTTCGCGGAAAGTGAACTTAGATTCGGTCCTGAATCGACT
## [4]    50 AACACGGTACTTCCACAGTCAACCCGCCGACTTGGAGAATTTAATCGACT
## [5]    50 TTAGCCGGGCGGTTATTCCCCTAGTGATCTTACTAAGATTTGCATCGACT
## [6]    50 AATACCTAAGCGAAGTGACAGATATGTTCGTCATTCATCCAGGATCGACT</code></pre>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="co">### PARTIAL ADAPTER IN 5'</span></span>
<span id="cb11-2"><a href="#cb11-2"></a>adapter &lt;-<span class="st"> "ATCGACT"</span></span>
<span id="cb11-3"><a href="#cb11-3"></a>subadapter &lt;-<span class="st"> </span><span class="kw">subseq</span>(adapter, <span class="dv">1</span>, <span class="dv">4</span>)</span>
<span id="cb11-4"><a href="#cb11-4"></a></span>
<span id="cb11-5"><a href="#cb11-5"></a><span class="co"># Create sequences with adapter</span></span>
<span id="cb11-6"><a href="#cb11-6"></a>my_seqs &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(input, subadapter)</span>
<span id="cb11-7"><a href="#cb11-7"></a>my_seqs &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html">DNAStringSet</a></span>(my_seqs)</span>
<span id="cb11-8"><a href="#cb11-8"></a>my_seqs</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 6
##     width seq
## [1]    47 TGGTCCGGTGTTCTGGCGGAATAGGTACAGTCCAGTAATTGCCATCG
## [2]    47 TCCCGCAGACGCTGGGTCCGGAATGCCCTTTCTGAGCAGCTCCATCG
## [3]    47 AGCCGTTTGACTTCGCGGAAAGTGAACTTAGATTCGGTCCTGAATCG
## [4]    47 AACACGGTACTTCCACAGTCAACCCGCCGACTTGGAGAATTTAATCG
## [5]    47 TTAGCCGGGCGGTTATTCCCCTAGTGATCTTACTAAGATTTGCATCG
## [6]    47 AATACCTAAGCGAAGTGACAGATATGTTCGTCATTCATCCAGGATCG</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="co"># create ShortReadQ object</span></span>
<span id="cb13-2"><a href="#cb13-2"></a>my_read &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ShortRead/man/ShortReadQ-class.html">ShortReadQ</a></span>(<span class="dt">sread =</span> my_seqs, <span class="dt">quality =</span> <span class="kw">subseq</span>(input_q, <span class="dv">1</span>, <span class="dv">47</span>), </span>
<span id="cb13-3"><a href="#cb13-3"></a><span class="dt">id =</span> input_names)</span>
<span id="cb13-4"><a href="#cb13-4"></a></span>
<span id="cb13-5"><a href="#cb13-5"></a><span class="co"># trim adapter</span></span>
<span id="cb13-6"><a href="#cb13-6"></a>filtered &lt;-<span class="st"> </span><span class="kw"><a href="../reference/adapter_filter.html">adapter_filter</a></span>(my_read, <span class="dt">Rpattern =</span> adapter)</span>
<span id="cb13-7"><a href="#cb13-7"></a><span class="kw"><a href="https://rdrr.io/pkg/ShortRead/man/ShortRead-class.html">sread</a></span>(filtered)</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 6
##     width seq
## [1]    43 TGGTCCGGTGTTCTGGCGGAATAGGTACAGTCCAGTAATTGCC
## [2]    43 TCCCGCAGACGCTGGGTCCGGAATGCCCTTTCTGAGCAGCTCC
## [3]    43 AGCCGTTTGACTTCGCGGAAAGTGAACTTAGATTCGGTCCTGA
## [4]    43 AACACGGTACTTCCACAGTCAACCCGCCGACTTGGAGAATTTA
## [5]    43 TTAGCCGGGCGGTTATTCCCCTAGTGATCTTACTAAGATTTGC
## [6]    43 AATACCTAAGCGAAGTGACAGATATGTTCGTCATTCATCCAGG</code></pre>
<p><a href="docs/reference/adapter_filter.html">Documentation of the function</a></p>
<p> </p>
<ul>
<li><strong>complex_filter</strong></li>
</ul>
<p>Removes low complexity sequences, computing the entropy with the dinucleotide frequency: <span class="math display">\[H_i = -\sum d_i * log_2(d_i)\]</span></p>
<p>where: <span class="math inline">\(d_i = D_i/ \sum_i^n D_i\)</span> represents the frequency of dinucleotides of the sequence <span class="math inline">\(i\)</span> relative to the frequency in the whole pool of sequences.</p>
<p>The relation <span class="math inline">\(H_i/H_r\)</span> between <span class="math inline">\(H_i\)</span> and a reference entropy value <span class="math inline">\(H_r\)</span> is computed, and the obtained relations are compared with a given complexity threshold. By default the program uses a reference entropy of 3.908, that corresponds to the entropy of the human genome in bits, and a complexity threshold of 0.5.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a><span class="co"># create  sequences of different width</span></span>
<span id="cb15-2"><a href="#cb15-2"></a><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">10</span>)</span>
<span id="cb15-3"><a href="#cb15-3"></a>input &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="dv">6</span>, <span class="dv">10</span>, <span class="dv">16</span>, <span class="dv">20</span>, <span class="dv">26</span>, <span class="dv">30</span>, <span class="dv">36</span>, <span class="dv">40</span>), </span>
<span id="cb15-4"><a href="#cb15-4"></a>            <span class="cf">function</span>(x) <span class="kw"><a href="../reference/random_seq.html">random_seq</a></span>(<span class="dv">1</span>, x))</span>
<span id="cb15-5"><a href="#cb15-5"></a></span>
<span id="cb15-6"><a href="#cb15-6"></a></span>
<span id="cb15-7"><a href="#cb15-7"></a><span class="co"># create repetitive "CG" sequences with length adequante </span></span>
<span id="cb15-8"><a href="#cb15-8"></a><span class="co"># for a total length input +  CG = 40</span></span>
<span id="cb15-9"><a href="#cb15-9"></a></span>
<span id="cb15-10"><a href="#cb15-10"></a>CG &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">20</span>, <span class="dv">17</span>, <span class="dv">15</span>, <span class="dv">12</span>, <span class="dv">10</span>, <span class="dv">7</span>, <span class="dv">5</span>, <span class="dv">2</span>, <span class="dv">0</span>), </span>
<span id="cb15-11"><a href="#cb15-11"></a>            <span class="cf">function</span>(x) <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">"CG"</span>, x), <span class="dt">collapse =</span> <span class="st">""</span>))</span>
<span id="cb15-12"><a href="#cb15-12"></a></span>
<span id="cb15-13"><a href="#cb15-13"></a><span class="co"># concatenate input and CG</span></span>
<span id="cb15-14"><a href="#cb15-14"></a>input  &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/mapply.html">mapply</a></span>(<span class="st">"paste"</span>, input, CG, <span class="dt">sep =</span> <span class="st">""</span>)</span>
<span id="cb15-15"><a href="#cb15-15"></a>input &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html">DNAStringSet</a></span>(input)</span>
<span id="cb15-16"><a href="#cb15-16"></a>input</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 9
##     width seq
## [1]    40 CGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCG
## [2]    40 TGGTCCCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCG
## [3]    40 GGTGTTCTGGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCG
## [4]    40 CGGAATAGGTACAGTCCGCGCGCGCGCGCGCGCGCGCGCG
## [5]    40 CAGTAATTGCCTCCCGCAGACGCGCGCGCGCGCGCGCGCG
## [6]    40 CGCTGGGTCCGGAATGCCCTTTCTGACGCGCGCGCGCGCG
## [7]    40 GCAGCTCCAGCCGTTTGACTTCGCGGAAAGCGCGCGCGCG
## [8]    40 TGAACTTAGATTCGGTCCTGAAACACGGTACTTCCACGCG
## [9]    40 CAGTCAACCCGCCGACTTGGAGAATTTATTAGCCGGGCGG</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a><span class="co"># plot relative entropy (E, Shannon 1948)</span></span>
<span id="cb17-2"><a href="#cb17-2"></a>H_plot &lt;-<span class="st"> </span><span class="cf">function</span>(x, <span class="dt">H_max =</span> <span class="fl">3.908135</span>) {</span>
<span id="cb17-3"><a href="#cb17-3"></a>    freq &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Biostrings/man/nucleotideFrequency.html">dinucleotideFrequency</a></span>(x)</span>
<span id="cb17-4"><a href="#cb17-4"></a>    freq  &lt;-<span class="st"> </span>freq <span class="op">/</span><span class="kw"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(freq)</span>
<span id="cb17-5"><a href="#cb17-5"></a>    H &lt;-<span class="st"> </span><span class="op">-</span><span class="kw"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(freq  <span class="op">*</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/Log.html">log2</a></span>(freq), <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)</span>
<span id="cb17-6"><a href="#cb17-6"></a>    <span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(H<span class="op">/</span>H_max, <span class="dt">type=</span><span class="st">"l"</span>, <span class="dt">xlab =</span> <span class="st">"Sequence"</span>, <span class="dt">ylab=</span> <span class="st">"E"</span>)</span>
<span id="cb17-7"><a href="#cb17-7"></a>    <span class="kw"><a href="https://rdrr.io/r/graphics/points.html">points</a></span>(H<span class="op">/</span>H_max, <span class="dt">col =</span> <span class="st">"#1a81c2"</span>, <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex =</span> <span class="dv">2</span>)</span>
<span id="cb17-8"><a href="#cb17-8"></a>}</span>
<span id="cb17-9"><a href="#cb17-9"></a></span>
<span id="cb17-10"><a href="#cb17-10"></a><span class="kw">H_plot</span>(input)</span></code></pre></div>
<p><img src="Overview_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p><strong>Figure 13</strong>: Relative entropy plot for the sequences before the operation</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="co"># create qualities of widths 40</span></span>
<span id="cb18-2"><a href="#cb18-2"></a><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">10</span>)</span>
<span id="cb18-3"><a href="#cb18-3"></a>input_q &lt;-<span class="st"> </span><span class="kw"><a href="../reference/random_qual.html">random_qual</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">30</span>,<span class="dv">40</span>), <span class="dt">slength =</span> <span class="dv">9</span>, <span class="dt">swidth =</span> <span class="dv">40</span>, </span>
<span id="cb18-4"><a href="#cb18-4"></a>    <span class="dt">encod =</span> <span class="st">"Sanger"</span>)</span>
<span id="cb18-5"><a href="#cb18-5"></a></span>
<span id="cb18-6"><a href="#cb18-6"></a><span class="co"># create names</span></span>
<span id="cb18-7"><a href="#cb18-7"></a>input_names &lt;-<span class="st"> </span><span class="kw"><a href="../reference/seq_names.html">seq_names</a></span>(<span class="dv">9</span>)</span>
<span id="cb18-8"><a href="#cb18-8"></a></span>
<span id="cb18-9"><a href="#cb18-9"></a></span>
<span id="cb18-10"><a href="#cb18-10"></a><span class="co"># create ShortReadQ object</span></span>
<span id="cb18-11"><a href="#cb18-11"></a>my_read &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ShortRead/man/ShortReadQ-class.html">ShortReadQ</a></span>(<span class="dt">sread =</span> input, <span class="dt">quality =</span> input_q, <span class="dt">id =</span> input_names)</span>
<span id="cb18-12"><a href="#cb18-12"></a></span>
<span id="cb18-13"><a href="#cb18-13"></a><span class="co"># apply the filter, </span></span>
<span id="cb18-14"><a href="#cb18-14"></a>filtered &lt;-<span class="st"> </span><span class="kw"><a href="../reference/complex_filter.html">complex_filter</a></span>(my_read)</span>
<span id="cb18-15"><a href="#cb18-15"></a><span class="kw"><a href="https://rdrr.io/pkg/ShortRead/man/ShortRead-class.html">sread</a></span>(filtered)</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 7
##     width seq
## [1]    40 GGTGTTCTGGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCG
## [2]    40 CGGAATAGGTACAGTCCGCGCGCGCGCGCGCGCGCGCGCG
## [3]    40 CAGTAATTGCCTCCCGCAGACGCGCGCGCGCGCGCGCGCG
## [4]    40 CGCTGGGTCCGGAATGCCCTTTCTGACGCGCGCGCGCGCG
## [5]    40 GCAGCTCCAGCCGTTTGACTTCGCGGAAAGCGCGCGCGCG
## [6]    40 TGAACTTAGATTCGGTCCTGAAACACGGTACTTCCACGCG
## [7]    40 CAGTCAACCCGCCGACTTGGAGAATTTATTAGCCGGGCGG</code></pre>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a><span class="kw">H_plot</span>(<span class="kw"><a href="https://rdrr.io/pkg/ShortRead/man/ShortRead-class.html">sread</a></span>(filtered))</span></code></pre></div>
<p><img src="Overview_files/figure-html/unnamed-chunk-8-1.png" width="700"><strong>Figure 14</strong>: Relative entropy plot for the sequences after the operation</p>
<p><a href="docs/reference/complex_filter.html">Documentation of the function</a></p>
<p> </p>
<ul>
<li><strong>fixed_filter</strong></li>
</ul>
<p>Removes the specified number of bases from 3’ or 5’.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a><span class="co"># create sequences, qualities and names of width 20</span></span>
<span id="cb21-2"><a href="#cb21-2"></a><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">10</span>)</span>
<span id="cb21-3"><a href="#cb21-3"></a>input &lt;-<span class="st"> </span><span class="kw"><a href="../reference/random_seq.html">random_seq</a></span>(<span class="dv">6</span>, <span class="dv">20</span>)</span>
<span id="cb21-4"><a href="#cb21-4"></a>input</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 6
##     width seq
## [1]    20 TGGTCCGGTGTTCTGGCGGA
## [2]    20 ATAGGTACAGTCCAGTAATT
## [3]    20 GCCTCCCGCAGACGCTGGGT
## [4]    20 CCGGAATGCCCTTTCTGAGC
## [5]    20 AGCTCCAGCCGTTTGACTTC
## [6]    20 GCGGAAAGTGAACTTAGATT</code></pre>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">10</span>)</span>
<span id="cb23-2"><a href="#cb23-2"></a>input_q &lt;-<span class="st"> </span><span class="kw"><a href="../reference/random_qual.html">random_qual</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">30</span>,<span class="dv">40</span>), <span class="dt">slength =</span> <span class="dv">6</span>, <span class="dt">swidth =</span> <span class="dv">20</span>, </span>
<span id="cb23-3"><a href="#cb23-3"></a>    <span class="dt">encod =</span> <span class="st">"Sanger"</span>)</span>
<span id="cb23-4"><a href="#cb23-4"></a></span>
<span id="cb23-5"><a href="#cb23-5"></a>input_names &lt;-<span class="st"> </span><span class="kw"><a href="../reference/seq_names.html">seq_names</a></span>(<span class="dv">6</span>)</span>
<span id="cb23-6"><a href="#cb23-6"></a></span>
<span id="cb23-7"><a href="#cb23-7"></a><span class="co"># create ShortReadQ object</span></span>
<span id="cb23-8"><a href="#cb23-8"></a>my_read &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ShortRead/man/ShortReadQ-class.html">ShortReadQ</a></span>(<span class="dt">sread =</span> input, <span class="dt">quality =</span> input_q, <span class="dt">id =</span> input_names)</span>
<span id="cb23-9"><a href="#cb23-9"></a></span>
<span id="cb23-10"><a href="#cb23-10"></a><span class="co"># apply the filter </span></span>
<span id="cb23-11"><a href="#cb23-11"></a>filtered3 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/fixed_filter.html">fixed_filter</a></span>(my_read, <span class="dt">trim5 =</span> <span class="dv">5</span>)</span>
<span id="cb23-12"><a href="#cb23-12"></a><span class="kw"><a href="https://rdrr.io/pkg/ShortRead/man/ShortRead-class.html">sread</a></span>(filtered3)</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 6
##     width seq
## [1]    15 TGGTCCGGTGTTCTG
## [2]    15 ATAGGTACAGTCCAG
## [3]    15 GCCTCCCGCAGACGC
## [4]    15 CCGGAATGCCCTTTC
## [5]    15 AGCTCCAGCCGTTTG
## [6]    15 GCGGAAAGTGAACTT</code></pre>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a>filtered5 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/fixed_filter.html">fixed_filter</a></span>(my_read, <span class="dt">trim3 =</span> <span class="dv">5</span>)</span>
<span id="cb25-2"><a href="#cb25-2"></a><span class="kw"><a href="https://rdrr.io/pkg/ShortRead/man/ShortRead-class.html">sread</a></span>(filtered5)</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 6
##     width seq
## [1]    15 CGGTGTTCTGGCGGA
## [2]    15 TACAGTCCAGTAATT
## [3]    15 CCGCAGACGCTGGGT
## [4]    15 ATGCCCTTTCTGAGC
## [5]    15 CAGCCGTTTGACTTC
## [6]    15 AAGTGAACTTAGATT</code></pre>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a>filtered3and5 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/fixed_filter.html">fixed_filter</a></span>(my_read, <span class="dt">trim3 =</span> <span class="dv">10</span>, <span class="dt">trim5 =</span> <span class="dv">5</span>)</span>
<span id="cb27-2"><a href="#cb27-2"></a><span class="kw"><a href="https://rdrr.io/pkg/ShortRead/man/ShortRead-class.html">sread</a></span>(filtered3and5)</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 6
##     width seq
## [1]     5 TTCTG
## [2]     5 TCCAG
## [3]     5 GACGC
## [4]     5 CTTTC
## [5]     5 GTTTG
## [6]     5 AACTT</code></pre>
<p><a href="docs/reference/fixed_filter.html">Documentation of the function</a></p>
<p> </p>
<ul>
<li><strong>length_filter</strong></li>
</ul>
<p>Removes sequences with a length lower than minimum threshold value or/and higher than a maximum threshold value.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a><span class="co"># create  ShortReadQ object width widths between 1 and 60</span></span>
<span id="cb29-2"><a href="#cb29-2"></a><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">10</span>)</span>
<span id="cb29-3"><a href="#cb29-3"></a>input &lt;-<span class="st"> </span><span class="kw"><a href="../reference/random_length.html">random_length</a></span>(<span class="dv">10</span>, <span class="dt">widths =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">60</span>)</span>
<span id="cb29-4"><a href="#cb29-4"></a><span class="kw"><a href="https://rdrr.io/pkg/ShortRead/man/ShortRead-class.html">sread</a></span>(input)</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 10
##      width seq
##  [1]    43 TCTGGCGGAATAGGTACAGTCCAGTAATTGCCTCCCGCAGACG
##  [2]     9 CTGGGTCCG
##  [3]    10 GAATGCCCTT
##  [4]    48 TCTGAGCAGCTCCAGCCGTTTGACTTCGCGGAAAGTGAACTTAGATTC
##  [5]    12 GGTCCTGAAACA
##  [6]    55 CGGTACTTCCACAGTCAACCCGCCGACTTGGAGAATTTATTAGCCGGGCGGTTAT
##  [7]     8 TCCCCTAG
##  [8]    54 TGATCTTACTAAGATTTGCAATACCTAAGCGAAGTGACAGATATGTTCGTCATT
##  [9]    39 CATCCAGGCAAGTGCGCGGACATCAATTACCACACAATT
## [10]    19 AAATATGACTCGCGTATCG</code></pre>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a><span class="co"># apply the filter, removing sequences with length &lt; 5 or length&gt; 30</span></span>
<span id="cb31-2"><a href="#cb31-2"></a>filtered &lt;-<span class="st"> </span><span class="kw"><a href="../reference/length_filter.html">length_filter</a></span>(input, <span class="dt">rm.min =</span> <span class="dv">5</span>, <span class="dt">rm.max =</span> <span class="dv">30</span>)</span>
<span id="cb31-3"><a href="#cb31-3"></a><span class="kw"><a href="https://rdrr.io/pkg/ShortRead/man/ShortRead-class.html">sread</a></span>(filtered)</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 5
##     width seq
## [1]     9 CTGGGTCCG
## [2]    10 GAATGCCCTT
## [3]    12 GGTCCTGAAACA
## [4]     8 TCCCCTAG
## [5]    19 AAATATGACTCGCGTATCG</code></pre>
<p><a href="docs/reference/length_filter.html">Documentation of the function</a></p>
<p> </p>
<ul>
<li><strong>n_filter</strong></li>
</ul>
<p>Wrapper of the <em><a href="https://bioconductor.org/packages/3.9/ShortRead">ShortRead</a></em> nFilter function. Removes all those sequences with a number of N’s &gt; a given threshold.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1"></a><span class="co"># create 10 sequences of width 20</span></span>
<span id="cb33-2"><a href="#cb33-2"></a><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">10</span>)</span>
<span id="cb33-3"><a href="#cb33-3"></a>input &lt;-<span class="st"> </span><span class="kw"><a href="../reference/random_seq.html">random_seq</a></span>(<span class="dv">10</span>, <span class="dv">20</span>)</span>
<span id="cb33-4"><a href="#cb33-4"></a>input</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 10
##      width seq
##  [1]    20 TGGTCCGGTGTTCTGGCGGA
##  [2]    20 ATAGGTACAGTCCAGTAATT
##  [3]    20 GCCTCCCGCAGACGCTGGGT
##  [4]    20 CCGGAATGCCCTTTCTGAGC
##  [5]    20 AGCTCCAGCCGTTTGACTTC
##  [6]    20 GCGGAAAGTGAACTTAGATT
##  [7]    20 CGGTCCTGAAACACGGTACT
##  [8]    20 TCCACAGTCAACCCGCCGAC
##  [9]    20 TTGGAGAATTTATTAGCCGG
## [10]    20 GCGGTTATTCCCCTAGTGAT</code></pre>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1"></a><span class="co"># inject N's</span></span>
<span id="cb35-2"><a href="#cb35-2"></a><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">10</span>)</span>
<span id="cb35-3"><a href="#cb35-3"></a>input &lt;-<span class="st"> </span><span class="kw"><a href="../reference/inject_letter_random.html">inject_letter_random</a></span>(input, <span class="dt">how_many_seqs =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,</span>
<span id="cb35-4"><a href="#cb35-4"></a>    <span class="dt">how_many =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>)</span>
<span id="cb35-5"><a href="#cb35-5"></a>input</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 10
##      width seq
##  [1]    20 TGGTCCGGTGTTCTGGCGGA
##  [2]    20 ATAGGTACAGTCCAGTAATT
##  [3]    20 GCCTCCCGCAGACGCTGGGT
##  [4]    20 CCGGAATGCCCTTTCTGAGC
##  [5]    20 AGCTCCAGCCGTTTGACTTC
##  [6]    20 GCGGAAAGTGAACTTAGATT
##  [7]    20 CNGTCNNNAANCNNNNTACN
##  [8]    20 NCCANANTCAACNCGCCNAC
##  [9]    20 TTGGAGNATNTATTNGCNNN
## [10]    20 GCGGTTATTCCCCTAGTGAT</code></pre>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1"></a><span class="co">#'  </span></span>
<span id="cb37-2"><a href="#cb37-2"></a><span class="kw"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/Biostrings/man/letterFrequency.html">letterFrequency</a></span>(input, <span class="st">"N"</span>), <span class="dt">breaks =</span> <span class="dv">0</span><span class="op">:</span><span class="dv">10</span>, </span>
<span id="cb37-3"><a href="#cb37-3"></a>    <span class="dt">main  =</span> <span class="st">"Ns Frequency"</span>, <span class="dt">xlab =</span> <span class="st">"# Ns"</span>,</span>
<span id="cb37-4"><a href="#cb37-4"></a>    <span class="dt">col =</span> <span class="st">"#1a81c2"</span>)</span></code></pre></div>
<p><img src="Overview_files/figure-html/unnamed-chunk-11-1.png" width="700"><strong>Figure 15</strong>: N’s histogram for the sequences before the filtering operation</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1"></a><span class="co"># Create qualities, names and ShortReadQ object</span></span>
<span id="cb38-2"><a href="#cb38-2"></a><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">10</span>)</span>
<span id="cb38-3"><a href="#cb38-3"></a>input_q &lt;-<span class="st"> </span><span class="kw"><a href="../reference/random_qual.html">random_qual</a></span>(<span class="dv">10</span>, <span class="dv">20</span>)</span>
<span id="cb38-4"><a href="#cb38-4"></a>input_names &lt;-<span class="st"> </span><span class="kw"><a href="../reference/seq_names.html">seq_names</a></span>(<span class="dv">10</span>)</span>
<span id="cb38-5"><a href="#cb38-5"></a>my_read &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ShortRead/man/ShortReadQ-class.html">ShortReadQ</a></span>(<span class="dt">sread =</span> input, <span class="dt">quality =</span> input_q, <span class="dt">id =</span> input_names)</span>
<span id="cb38-6"><a href="#cb38-6"></a></span>
<span id="cb38-7"><a href="#cb38-7"></a><span class="co"># Apply the filter </span></span>
<span id="cb38-8"><a href="#cb38-8"></a>filtered &lt;-<span class="st"> </span><span class="kw"><a href="../reference/n_filter.html">n_filter</a></span>(my_read, <span class="dt">rm.N =</span> <span class="dv">3</span>)</span>
<span id="cb38-9"><a href="#cb38-9"></a><span class="kw"><a href="https://rdrr.io/pkg/ShortRead/man/ShortRead-class.html">sread</a></span>(filtered)</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 7
##     width seq
## [1]    20 TGGTCCGGTGTTCTGGCGGA
## [2]    20 ATAGGTACAGTCCAGTAATT
## [3]    20 GCCTCCCGCAGACGCTGGGT
## [4]    20 CCGGAATGCCCTTTCTGAGC
## [5]    20 AGCTCCAGCCGTTTGACTTC
## [6]    20 GCGGAAAGTGAACTTAGATT
## [7]    20 GCGGTTATTCCCCTAGTGAT</code></pre>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1"></a><span class="kw"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/Biostrings/man/letterFrequency.html">letterFrequency</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/ShortRead/man/ShortRead-class.html">sread</a></span>(filtered), <span class="st">"N"</span>), </span>
<span id="cb40-2"><a href="#cb40-2"></a>    <span class="dt">main =</span> <span class="st">"Ns distribution"</span>, <span class="dt">xlab =</span> <span class="st">""</span>,</span>
<span id="cb40-3"><a href="#cb40-3"></a>    <span class="dt">col =</span> <span class="st">"#1a81c2"</span>)</span></code></pre></div>
<p><img src="Overview_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p><strong>Figure 16</strong>: N’s histogram for the sequences after the filtering operation</p>
<p><a href="docs/reference/n_filter.html">Documentation of the function</a></p>
<p> </p>
<ul>
<li><strong>qmean_filter</strong></li>
</ul>
<p>Removes those sequences with quality &lt; a give threshold.</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1"></a><span class="co"># create 30 sequences of width 20, 15 with low quality and 15 with high quality</span></span>
<span id="cb41-2"><a href="#cb41-2"></a><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">10</span>)</span>
<span id="cb41-3"><a href="#cb41-3"></a>input &lt;-<span class="st"> </span><span class="kw"><a href="../reference/random_seq.html">random_seq</a></span>(<span class="dv">30</span>, <span class="dv">20</span>)</span>
<span id="cb41-4"><a href="#cb41-4"></a></span>
<span id="cb41-5"><a href="#cb41-5"></a><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">10</span>)</span>
<span id="cb41-6"><a href="#cb41-6"></a>my_qual_H &lt;-<span class="st"> </span><span class="kw"><a href="../reference/random_qual.html">random_qual</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">30</span>,<span class="dv">40</span>), <span class="dt">slength =</span> <span class="dv">15</span>, <span class="dt">swidth =</span> <span class="dv">20</span>,</span>
<span id="cb41-7"><a href="#cb41-7"></a>    <span class="dt">encod =</span> <span class="st">"Sanger"</span>)</span>
<span id="cb41-8"><a href="#cb41-8"></a></span>
<span id="cb41-9"><a href="#cb41-9"></a><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">10</span>)</span>
<span id="cb41-10"><a href="#cb41-10"></a>my_qual_L &lt;-<span class="st"> </span><span class="kw"><a href="../reference/random_qual.html">random_qual</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">5</span>,<span class="dv">30</span>), <span class="dt">slength =</span> <span class="dv">15</span>, <span class="dt">swidth =</span> <span class="dv">20</span>, </span>
<span id="cb41-11"><a href="#cb41-11"></a>    <span class="dt">encod =</span> <span class="st">"Sanger"</span>)</span>
<span id="cb41-12"><a href="#cb41-12"></a>input_q&lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(my_qual_H, my_qual_L)</span>
<span id="cb41-13"><a href="#cb41-13"></a></span>
<span id="cb41-14"><a href="#cb41-14"></a>input_names &lt;-<span class="st"> </span><span class="kw"><a href="../reference/seq_names.html">seq_names</a></span>(<span class="dv">30</span>)</span>
<span id="cb41-15"><a href="#cb41-15"></a>my_read &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ShortRead/man/ShortReadQ-class.html">ShortReadQ</a></span>(<span class="dt">sread =</span> input, <span class="dt">quality =</span> input_q, <span class="dt">id =</span> input_names)</span>
<span id="cb41-16"><a href="#cb41-16"></a></span>
<span id="cb41-17"><a href="#cb41-17"></a><span class="co"># Plot of average qualities</span></span>
<span id="cb41-18"><a href="#cb41-18"></a>qual_plot &lt;-<span class="st"> </span><span class="cf">function</span>(x, cutoff) {</span>
<span id="cb41-19"><a href="#cb41-19"></a>q &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ShortRead/man/alphabetScore.html">alphabetScore</a></span>(x) <span class="op">/</span><span class="st"> </span><span class="kw">width</span>(x)</span>
<span id="cb41-20"><a href="#cb41-20"></a><span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(q, <span class="dt">type=</span><span class="st">"l"</span>, <span class="dt">xlab =</span> <span class="st">"Sequence"</span>, <span class="dt">ylab=</span> <span class="st">"Average quality"</span>, <span class="dt">ylim =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>, <span class="dv">40</span>))</span>
<span id="cb41-21"><a href="#cb41-21"></a><span class="kw"><a href="https://rdrr.io/r/graphics/points.html">points</a></span>(q, <span class="dt">col =</span> <span class="st">"#1a81c2"</span>, <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex =</span> <span class="dv">2</span>)</span>
<span id="cb41-22"><a href="#cb41-22"></a><span class="kw"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span>(q), <span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(cutoff, <span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(q)), <span class="dt">type=</span><span class="st">"l"</span>, <span class="dt">col =</span> <span class="st">"red"</span>, <span class="dt">lty=</span><span class="dv">2</span>)</span>
<span id="cb41-23"><a href="#cb41-23"></a><span class="kw"><a href="https://rdrr.io/r/graphics/text.html">text</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(q), cutoff<span class="op">+</span><span class="dv">2</span>, cutoff)</span>
<span id="cb41-24"><a href="#cb41-24"></a>}</span>
<span id="cb41-25"><a href="#cb41-25"></a></span>
<span id="cb41-26"><a href="#cb41-26"></a><span class="co">#' Average qualities before</span></span>
<span id="cb41-27"><a href="#cb41-27"></a><span class="kw">qual_plot</span>(my_read, <span class="dt">cutoff =</span> <span class="dv">30</span>)</span></code></pre></div>
<p><img src="Overview_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p><strong>Figure 17</strong>: Average qualities before the filtering operation</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1"></a><span class="co"># Apply the filter</span></span>
<span id="cb42-2"><a href="#cb42-2"></a>filtered &lt;-<span class="st"> </span><span class="kw"><a href="../reference/qmean_filter.html">qmean_filter</a></span>(my_read, <span class="dt">minq =</span> <span class="dv">30</span>)</span>
<span id="cb42-3"><a href="#cb42-3"></a></span>
<span id="cb42-4"><a href="#cb42-4"></a><span class="co"># Average qualities after</span></span>
<span id="cb42-5"><a href="#cb42-5"></a><span class="kw">qual_plot</span>(filtered, <span class="dt">cutoff =</span> <span class="dv">30</span>)</span></code></pre></div>
<p><img src="Overview_files/figure-html/unnamed-chunk-14-1.png" width="700"><strong>Figure 18</strong>: Average qualities after the filtering operation</p>
<p><a href="docs/reference/qmean_filter.html">Documentation of the function</a></p>
<p> </p>
<ul>
<li><strong>seq_filter</strong></li>
</ul>
<p>Removes sequences that match those passed as argument.</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1"></a><span class="co"># Generate random sequences</span></span>
<span id="cb43-2"><a href="#cb43-2"></a><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">10</span>)</span>
<span id="cb43-3"><a href="#cb43-3"></a>input &lt;-<span class="st"> </span><span class="kw"><a href="../reference/random_length.html">random_length</a></span>(<span class="dv">30</span>, <span class="dv">3</span><span class="op">:</span><span class="dv">7</span>)</span>
<span id="cb43-4"><a href="#cb43-4"></a></span>
<span id="cb43-5"><a href="#cb43-5"></a><span class="co"># Remove sequences that contain the following patterns:</span></span>
<span id="cb43-6"><a href="#cb43-6"></a>rm.seq  =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"TGGTC"</span>, <span class="st">"CGGT"</span>, <span class="st">"GTTCT"</span>, <span class="st">"ATA"</span>)</span>
<span id="cb43-7"><a href="#cb43-7"></a>match_before &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(rm.seq, <span class="cf">function</span>(x) <span class="kw"><a href="https://rdrr.io/r/base/grep.html">grep</a></span>(x, </span>
<span id="cb43-8"><a href="#cb43-8"></a><span class="kw"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/ShortRead/man/ShortRead-class.html">sread</a></span>(input)))))</span>
<span id="cb43-9"><a href="#cb43-9"></a>match_before</span></code></pre></div>
<pre><code>## [1] 14 17 27</code></pre>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1"></a>filtered &lt;-<span class="st"> </span><span class="kw"><a href="../reference/seq_filter.html">seq_filter</a></span>(input,<span class="dt">rm.seq =</span>  rm.seq)</span>
<span id="cb45-2"><a href="#cb45-2"></a></span>
<span id="cb45-3"><a href="#cb45-3"></a><span class="co"># Verify that matching sequences were removed</span></span>
<span id="cb45-4"><a href="#cb45-4"></a>match_after &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(rm.seq, <span class="cf">function</span>(x) {</span>
<span id="cb45-5"><a href="#cb45-5"></a>                <span class="kw"><a href="https://rdrr.io/r/base/grep.html">grep</a></span>(x, <span class="kw"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/ShortRead/man/ShortRead-class.html">sread</a></span>(filtered)))}))</span>
<span id="cb45-6"><a href="#cb45-6"></a>match_after</span></code></pre></div>
<pre><code>## [1] 14 26</code></pre>
<p><a href="docs/reference/seq_filter.html">Documentation of the function</a></p>
<p> </p>
<ul>
<li><strong>trim3q_filter</strong></li>
</ul>
<p>Removes from the 3’ ends in-tandem nucleotides with a quality &lt; a threshold value.</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1"></a><span class="co"># Create 6 sequences of width 20</span></span>
<span id="cb47-2"><a href="#cb47-2"></a><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">10</span>)</span>
<span id="cb47-3"><a href="#cb47-3"></a>input &lt;-<span class="st"> </span><span class="kw"><a href="../reference/random_seq.html">random_seq</a></span>(<span class="dv">6</span>, <span class="dv">20</span>)</span>
<span id="cb47-4"><a href="#cb47-4"></a>input</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 6
##     width seq
## [1]    20 TGGTCCGGTGTTCTGGCGGA
## [2]    20 ATAGGTACAGTCCAGTAATT
## [3]    20 GCCTCCCGCAGACGCTGGGT
## [4]    20 CCGGAATGCCCTTTCTGAGC
## [5]    20 AGCTCCAGCCGTTTGACTTC
## [6]    20 GCGGAAAGTGAACTTAGATT</code></pre>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1"></a><span class="co"># Create Phred+33 qualities of width 15 and paste to qualities of length </span></span>
<span id="cb49-2"><a href="#cb49-2"></a><span class="co"># 5 used for the tails.</span></span>
<span id="cb49-3"><a href="#cb49-3"></a><span class="co"># for three of the sequences, put low qualities in tails</span></span>
<span id="cb49-4"><a href="#cb49-4"></a><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">10</span>)</span>
<span id="cb49-5"><a href="#cb49-5"></a>my_qual &lt;-<span class="st"> </span><span class="kw"><a href="../reference/random_qual.html">random_qual</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">30</span>,<span class="dv">40</span>), <span class="dt">slength =</span> <span class="dv">6</span>, <span class="dt">swidth =</span> <span class="dv">15</span>, </span>
<span id="cb49-6"><a href="#cb49-6"></a>    <span class="dt">encod =</span> <span class="st">"Sanger"</span>)</span>
<span id="cb49-7"><a href="#cb49-7"></a><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">10</span>)</span>
<span id="cb49-8"><a href="#cb49-8"></a>tails &lt;-<span class="st">   </span><span class="kw"><a href="../reference/random_qual.html">random_qual</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">30</span>,<span class="dv">40</span>), <span class="dt">slength =</span> <span class="dv">6</span>, <span class="dt">swidth =</span> <span class="dv">5</span>, </span>
<span id="cb49-9"><a href="#cb49-9"></a>    <span class="dt">encod =</span> <span class="st">"Sanger"</span>)</span>
<span id="cb49-10"><a href="#cb49-10"></a></span>
<span id="cb49-11"><a href="#cb49-11"></a><span class="co"># Low quality tails in sequences 2, 3 &amp; 4</span></span>
<span id="cb49-12"><a href="#cb49-12"></a><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">10</span>)</span>
<span id="cb49-13"><a href="#cb49-13"></a>tails[<span class="dv">2</span><span class="op">:</span><span class="dv">4</span>] &lt;-<span class="st"> </span><span class="kw"><a href="../reference/random_qual.html">random_qual</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">3</span>, <span class="dv">20</span>), <span class="dt">slength =</span> <span class="dv">3</span>, <span class="dt">swidth =</span> <span class="dv">5</span>,</span>
<span id="cb49-14"><a href="#cb49-14"></a>    <span class="dt">encod =</span> <span class="st">"Sanger"</span>)</span>
<span id="cb49-15"><a href="#cb49-15"></a>my_qual &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(my_qual, tails)</span>
<span id="cb49-16"><a href="#cb49-16"></a>input_q &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html">BStringSet</a></span>(my_qual)</span>
<span id="cb49-17"><a href="#cb49-17"></a>input_q</span></code></pre></div>
<pre><code>##   A BStringSet instance of length 6
##     width seq
## [1]    20 EGFEDIBEH@C@DD?EAAID
## [2]    20 I?EGDHIBEG?BHFG,%),4
## [3]    20 ACCFBBFCI?I@HBC402+,
## [4]    20 CGIAFGB@?AIDF@I14)2+
## [5]    20 IB@ACAAC?AGEDDHC?BEB
## [6]    20 BH?GFFHHG?DABECFEEDE</code></pre>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1"></a><span class="co"># Watch qualities before filtering</span></span>
<span id="cb51-2"><a href="#cb51-2"></a><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/Biostrings/man/XStringQuality-class.html">PhredQuality</a></span>(input_q))</span></code></pre></div>
<pre><code>##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13] [,14]
## [1,]   36   38   37   36   35   40   33   36   39    31    34    31    35    35
## [2,]   40   30   36   38   35   39   40   33   36    38    30    33    39    37
## [3,]   32   34   34   37   33   33   37   34   40    30    40    31    39    33
## [4,]   34   38   40   32   37   38   33   31   30    32    40    35    37    31
## [5,]   40   33   31   32   34   32   32   34   30    32    38    36    35    35
## [6,]   33   39   30   38   37   37   39   39   38    30    35    32    33    36
##      [,15] [,16] [,17] [,18] [,19] [,20]
## [1,]    30    36    32    32    40    35
## [2,]    38    11     4     8    11    19
## [3,]    34    19    15    17    10    11
## [4,]    40    16    19     8    17    10
## [5,]    39    34    30    33    36    33
## [6,]    34    37    36    36    35    36</code></pre>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1"></a><span class="co"># Create names and ShortReadQ object</span></span>
<span id="cb53-2"><a href="#cb53-2"></a>input_names &lt;-<span class="st"> </span><span class="kw"><a href="../reference/seq_names.html">seq_names</a></span>(<span class="dv">6</span>)</span>
<span id="cb53-3"><a href="#cb53-3"></a>my_read &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ShortRead/man/ShortReadQ-class.html">ShortReadQ</a></span>(<span class="dt">sread =</span> input, <span class="dt">quality =</span> input_q, <span class="dt">id =</span> input_names)</span>
<span id="cb53-4"><a href="#cb53-4"></a></span>
<span id="cb53-5"><a href="#cb53-5"></a><span class="co"># Apply the filter </span></span>
<span id="cb53-6"><a href="#cb53-6"></a>filtered &lt;-<span class="st"> </span><span class="kw"><a href="../reference/trim3q_filter.html">trim3q_filter</a></span>(my_read, <span class="dt">rm.3qual =</span> <span class="dv">28</span>)</span>
<span id="cb53-7"><a href="#cb53-7"></a><span class="kw"><a href="https://rdrr.io/pkg/ShortRead/man/ShortRead-class.html">sread</a></span>(filtered)</span></code></pre></div>
<pre><code>##   A DNAStringSet instance of length 6
##     width seq
## [1]    20 TGGTCCGGTGTTCTGGCGGA
## [2]    15 ATAGGTACAGTCCAG
## [3]    15 GCCTCCCGCAGACGC
## [4]    15 CCGGAATGCCCTTTC
## [5]    20 AGCTCCAGCCGTTTGACTTC
## [6]    20 GCGGAAAGTGAACTTAGATT</code></pre>
<p><a href="docs/reference/trim3q_filter.html">Documentation of the function</a></p>
<p> </p>
<ul>
<li><strong>unique_filter</strong></li>
</ul>
<p>Wrapper of the <em><a href="https://bioconductor.org/packages/3.9/ShortRead">ShortRead</a></em> occurrenceFilter function. that removes duplicated sequences.</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1"></a><span class="co"># Create duplicated sequences</span></span>
<span id="cb55-2"><a href="#cb55-2"></a>s &lt;-<span class="st"> </span><span class="kw"><a href="../reference/random_seq.html">random_seq</a></span>(<span class="dv">10</span>, <span class="dv">10</span>)</span>
<span id="cb55-3"><a href="#cb55-3"></a>s &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(s, <span class="dv">30</span>, <span class="dt">replace =</span> <span class="ot">TRUE</span>)</span>
<span id="cb55-4"><a href="#cb55-4"></a></span>
<span id="cb55-5"><a href="#cb55-5"></a><span class="co"># Create a ShortReadQ object</span></span>
<span id="cb55-6"><a href="#cb55-6"></a>q &lt;-<span class="st"> </span><span class="kw"><a href="../reference/random_qual.html">random_qual</a></span>(<span class="dv">30</span>, <span class="dv">10</span>)</span>
<span id="cb55-7"><a href="#cb55-7"></a>n &lt;-<span class="st"> </span><span class="kw"><a href="../reference/seq_names.html">seq_names</a></span>(<span class="dv">30</span>)</span>
<span id="cb55-8"><a href="#cb55-8"></a>my_read &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ShortRead/man/ShortReadQ-class.html">ShortReadQ</a></span>(<span class="dt">sread =</span> s, <span class="dt">quality =</span> q, <span class="dt">id =</span> n)</span>
<span id="cb55-9"><a href="#cb55-9"></a></span>
<span id="cb55-10"><a href="#cb55-10"></a><span class="co"># Check presence of duplicates</span></span>
<span id="cb55-11"><a href="#cb55-11"></a><span class="kw">isUnique</span>(<span class="kw"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/ShortRead/man/ShortRead-class.html">sread</a></span>(my_read)))</span></code></pre></div>
<pre><code>##  [1] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
## [13] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
## [25] FALSE FALSE FALSE FALSE FALSE  TRUE</code></pre>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1"></a><span class="co"># Apply the filter</span></span>
<span id="cb57-2"><a href="#cb57-2"></a>filtered &lt;-<span class="st"> </span><span class="kw"><a href="../reference/unique_filter.html">unique_filter</a></span>(my_read)</span>
<span id="cb57-3"><a href="#cb57-3"></a><span class="kw">isUnique</span>(<span class="kw"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/ShortRead/man/ShortRead-class.html">sread</a></span>(filtered)))</span></code></pre></div>
<pre><code>##  [1] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</code></pre>
<p><a href="docs/reference/unique_filter.html">Documentation of the function</a></p>
<p> </p>
</div>
<div id="auxiliary-functions" class="section level2">
<h2 class="hasAnchor">
<a href="#auxiliary-functions" class="anchor"></a>Auxiliary functions</h2>
<ul>
<li><strong>random_seq</strong></li>
</ul>
<p>Create a vector of random sequences, for a set of specificied parameters.</p>
<p><a href="docs/reference/random_seq.html">Documentation of the function</a></p>
<p> </p>
<ul>
<li><strong>random_qual</strong></li>
</ul>
<p>Create a vector of random qualities for a given encoding and a set of specified parameters.</p>
<p><a href="docs/reference/random_qual.html">Documentation of the function</a></p>
<p> </p>
<ul>
<li><strong>seq_names</strong></li>
</ul>
<p>Create a vector of names for a set of sequences.</p>
<p><a href="docs/reference/seq_names.html">Documentation of the function</a></p>
<p> </p>
<ul>
<li><strong>random_length</strong></li>
</ul>
<p>Create a set of sequences with random lengths.</p>
<p><a href="docs/reference/seq_names.html">Documentation of the function</a></p>
<p> </p>
<ul>
<li><strong>inject_letter_random</strong></li>
</ul>
<p>Inject a character (e.g., ‘N’) at random positions, given a set of parameters.</p>
<p><a href="docs/reference/inject_letter_random.html">Documentation of the function</a></p>
<p> </p>
<ul>
<li><strong>check_encoding</strong></li>
</ul>
<p>The function allows to check quality encoding. It detects encodings with the following formats:</p>
<table class="table">
<thead><tr class="header">
<th>Format</th>
<th>Expected range</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Sanger</td>
<td>[0, 40]</td>
</tr>
<tr class="even">
<td>Illumina 1.8</td>
<td>[0, 41]</td>
</tr>
<tr class="odd">
<td>Illumina 1.5</td>
<td>[0, 40]</td>
</tr>
<tr class="even">
<td>Illumina 1.3</td>
<td>[3, 40]</td>
</tr>
<tr class="odd">
<td>Solexa</td>
<td>[-5, 40]</td>
</tr>
</tbody>
</table>
<p><a href="docs/reference/check_encoding.html">Documentation of the function</a></p>
<p> </p>
</div>
</div>
<div id="contact-information" class="section level1">
<h1 class="hasAnchor">
<a href="#contact-information" class="anchor"></a>Contact information</h1>
<p>Mantainer: Leandro Roser - <a href="mailto:learoser@gmail.com">learoser@gmail.com</a></p>
<p> </p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#launching-the-application">Launching the application</a></li>
      <li>
<a href="#description-of-the-application">Description of the application</a><ul class="nav nav-pills nav-stacked">
<li><a href="#first-panel">First panel</a></li>
      <li><a href="#second-panel">Second panel</a></li>
      <li><a href="#third-panel">Third panel</a></li>
      </ul>
</li>
      <li><a href="#a-worked-example-fastq-processing-in-a-nutshell">A worked example: FASTQ processing in a nutshell</a></li>
      <li>
<a href="#advanced-use-of-the-package">Advanced use of the package</a><ul class="nav nav-pills nav-stacked">
<li><a href="#main-functions">Main functions</a></li>
      <li><a href="#auxiliary-functions">Auxiliary functions</a></li>
      </ul>
</li>
      <li><a href="#contact-information">Contact information</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Leandro Roser, Fernán Agüero, Daniel Sánchez.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
