<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Remove full and partial adapters from a ShortReadQ object — adapter_filter • FastqCleaner</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">FastqCleaner</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Overview.html">An Introduction to FastqCleaner</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Remove full and partial adapters from a ShortReadQ object</h1>
    </div>

    
    <p>This program can remove adapters and partial 
adapters from 3' and 5', using the
<code><a href='http://www.rdocumentation.org/packages/Biostrings/topics/lowlevel-matching'>isMatchingEndingAt</a></code> and 
<code><a href='http://www.rdocumentation.org/packages/Biostrings/topics/lowlevel-matching'>isMatchingStartingAt</a></code> 
of <span class="pkg">Biostrings</span>.  The program extends the methodology of
the <code><a href='http://www.rdocumentation.org/packages/Biostrings/topics/trimLRPatterns'>trimLRPatterns</a></code> function of <span class="pkg">Biostrings</span>
and can in addition remove adapters present within reads. 
For a given position in the read, these two functions return TRUE 
when a match is present between a substring of the read and the adapter.
As <code><a href='http://www.rdocumentation.org/packages/Biostrings/topics/trimLRPatterns'>trimLRPatterns</a></code>, the function also selects
the best match as the longest subsequence that includes the aligned
region and goes up to the end of the sequence in the corresponding flank.
If several valid matches are found, the function removes the
largest subsequence.
Adapters can be anchored or not.  Two methods are available: 
one based on the exact matching of the adapter and
the reads, and other in an error rate. When indels are allowed, 
the second method uses the 'edit distance' between the sequences 
and the adapter.</p>
    

    <pre class="usage"><span class='fu'>adapter_filter</span>(<span class='no'>input</span>, <span class='kw'>Lpattern</span> <span class='kw'>=</span> <span class='st'>""</span>, <span class='kw'>Rpattern</span> <span class='kw'>=</span> <span class='st'>""</span>, <span class='kw'>method</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"exact"</span>,
<span class='st'>"er"</span>), <span class='kw'>rc.L</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>rc.R</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>first</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"R"</span>, <span class='st'>"L"</span>),
<span class='kw'>with_indels</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>error_rate</span> <span class='kw'>=</span> <span class='fl'>0.2</span>, <span class='kw'>anchored</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
<span class='kw'>fixed</span> <span class='kw'>=</span> <span class='st'>"subject"</span>, <span class='kw'>remove_zero</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>checks</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
<span class='kw'>min_match_flank</span> <span class='kw'>=</span> <span class='fl'>1L</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>input</th>
      <td><p><code><a href='http://www.rdocumentation.org/packages/ShortRead/topics/ShortReadQ-class'>ShortReadQ</a></code> object</p></td>
    </tr>
    <tr>
      <th>Lpattern</th>
      <td><p>5' pattern (character or
<code><a href='http://www.rdocumentation.org/packages/Biostrings/topics/DNAString-class'>DNAString</a></code> object)</p></td>
    </tr>
    <tr>
      <th>Rpattern</th>
      <td><p>3' pattern (character or 
<code><a href='http://www.rdocumentation.org/packages/Biostrings/topics/DNAString-class'>DNAString</a></code> object)</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>Method used for trimming. If 'exact' the metod is based
on the exact matching of a subsequences of the subject and adapter(s).
If 'er' the metod is based on the error-rate between the subsequences, 
allowing mismatches in any place.</p></td>
    </tr>
    <tr>
      <th>rc.L</th>
      <td><p>reverse complement Lpattern? default FALSE</p></td>
    </tr>
    <tr>
      <th>rc.R</th>
      <td><p>reverse complement Rpatter? default FALSE</p></td>
    </tr>
    <tr>
      <th>first</th>
      <td><p>trim first right('R') or left ('L') side of sequences
when both Lpattern and Rpattern are passed.</p></td>
    </tr>
    <tr>
      <th>with_indels</th>
      <td><p>allow indels? This is available only 
when error_rate is not null.</p></td>
    </tr>
    <tr>
      <th>error_rate</th>
      <td><p>Error rate (value in [0, 1] value for 'er' method.
The error rate is the proportion of mismatches allowed between
the adapter and the aligned portion  of the subject.
For a given adapter A, the number of allowed mismatches between each 
subsequence s of A and the subject is computed as error_rate * L_s,
where L_s is the length of the subsequence s.</p></td>
    </tr>
    <tr>
      <th>anchored</th>
      <td><p>Adapter or partial adapter within sequence
(anchored = FALSE, default)
or only in 3' and 5' terminals (anchored = TRUE).</p></td>
    </tr>
    <tr>
      <th>fixed</th>
      <td><p>Parameter passed to 
<code><a href='http://www.rdocumentation.org/packages/Biostrings/topics/lowlevel-matching'>isMatchingStartingAt</a></code> 
or <code><a href='http://www.rdocumentation.org/packages/Biostrings/topics/lowlevel-matching'>isMatchingEndingAt</a></code>. 
Default 'subject', where only ambiguities in the pattern are
interpreted as wildcard.</p></td>
    </tr>
    <tr>
      <th>remove_zero</th>
      <td><p>Remove zero-length sequences? Default TRUE.</p></td>
    </tr>
    <tr>
      <th>checks</th>
      <td><p>perform checks? Default TRUE.</p></td>
    </tr>
    <tr>
      <th>min_match_flank</th>
      <td><p>When a match is found between a subsequence
of the subject and the adapter in the corresponding flank,  
which would be the minimum length of the overlapping region 
(threshold) form trimming?  Default is 0L (trim when &gt;= 1 
base(s) match).</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>additional parameters passed to
<code><a href='http://www.rdocumentation.org/packages/Biostrings/topics/lowlevel-matching'>isMatchingStartingAt</a></code> 
or <code><a href='http://www.rdocumentation.org/packages/Biostrings/topics/lowlevel-matching'>isMatchingEndingAt</a></code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Filtered <code><a href='http://www.rdocumentation.org/packages/ShortRead/topics/ShortReadQ-class'>ShortReadQ</a></code> 
object</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='fu'>require</span>(<span class='st'>'Biostrings'</span>)</div><div class='output co'>#&gt; <span class='message'>Loading required package: Biostrings</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: BiocGenerics</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: parallel</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘BiocGenerics’</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:parallel’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,</span>
#&gt; <span class='message'>    clusterExport, clusterMap, parApply, parCapply, parLapply,</span>
#&gt; <span class='message'>    parLapplyLB, parRapply, parSapply, parSapplyLB</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:stats’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    IQR, mad, sd, var, xtabs</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:base’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    anyDuplicated, append, as.data.frame, cbind, colMeans, colnames,</span>
#&gt; <span class='message'>    colSums, do.call, duplicated, eval, evalq, Filter, Find, get, grep,</span>
#&gt; <span class='message'>    grepl, intersect, is.unsorted, lapply, lengths, Map, mapply, match,</span>
#&gt; <span class='message'>    mget, order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,</span>
#&gt; <span class='message'>    rbind, Reduce, rowMeans, rownames, rowSums, sapply, setdiff, sort,</span>
#&gt; <span class='message'>    table, tapply, union, unique, unsplit, which, which.max, which.min</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: S4Vectors</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: stats4</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘S4Vectors’</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:base’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    expand.grid</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: IRanges</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: XVector</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘Biostrings’</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:base’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    strsplit</span></div><div class='input'><span class='fu'>require</span>(<span class='st'>'ShortRead'</span>)</div><div class='output co'>#&gt; <span class='message'>Loading required package: ShortRead</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: BiocParallel</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: Rsamtools</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: GenomeInfoDb</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: GenomicRanges</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: GenomicAlignments</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: SummarizedExperiment</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: Biobase</span></div><div class='output co'>#&gt; <span class='message'>Welcome to Bioconductor</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    Vignettes contain introductory material; view with</span>
#&gt; <span class='message'>    'browseVignettes()'. To cite Bioconductor, see</span>
#&gt; <span class='message'>    'citation("Biobase")', and for packages 'citation("pkgname")'.</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: DelayedArray</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: matrixStats</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘matrixStats’</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:Biobase’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    anyMissing, rowMedians</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘DelayedArray’</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:matrixStats’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:Biostrings’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    type</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:base’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    apply</span></div><div class='input'>
<span class='co'># create 6 sequences of width 43</span>
<span class='no'>input</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='random_seq.html'>random_seq</a></span>(<span class='fl'>6</span>, <span class='fl'>43</span>, <span class='kw'>seed</span> <span class='kw'>=</span> <span class='fl'>10</span>)

<span class='co'># add adapter in 3' reverse complemented. In read 1,</span>
<span class='co'># it will appear the 5' adapter of read 2 reverse complemented.</span>
<span class='no'>adapter</span> <span class='kw'>&lt;-</span> <span class='st'>'ATCGACT'</span>

<span class='no'>input</span> <span class='kw'>&lt;-</span> <span class='fu'>paste0</span>(<span class='no'>input</span>, <span class='fu'>as.character</span>(<span class='fu'><a href='http://www.rdocumentation.org/packages/Biostrings/topics/reverseComplement'>reverseComplement</a></span>(<span class='fu'><a href='http://www.rdocumentation.org/packages/Biostrings/topics/DNAString-class'>DNAString</a></span>(<span class='no'>adapter</span>))))
<span class='no'>input</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='http://www.rdocumentation.org/packages/Biostrings/topics/XStringSet-class'>DNAStringSet</a></span>(<span class='no'>input</span>)

<span class='co'># create qualities of width 50</span>
<span class='no'>input_q</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='random_qual.html'>random_qual</a></span>(<span class='fu'>c</span>(<span class='fl'>30</span>,<span class='fl'>40</span>), <span class='kw'>slength</span> <span class='kw'>=</span> <span class='fl'>6</span>, <span class='kw'>swidth</span> <span class='kw'>=</span> <span class='fl'>50</span>,
<span class='kw'>seed</span> <span class='kw'>=</span> <span class='fl'>10</span>, <span class='kw'>encod</span> <span class='kw'>=</span> <span class='st'>'Sanger'</span>)

<span class='co'># create names</span>
<span class='no'>input_names</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='seq_names.html'>seq_names</a></span>(<span class='fu'>length</span>(<span class='no'>input</span>))

<span class='co'># create ShortReadQ object</span>
<span class='no'>my_read</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='http://www.rdocumentation.org/packages/ShortRead/topics/ShortReadQ-class'>ShortReadQ</a></span>(<span class='kw'>sread</span> <span class='kw'>=</span> <span class='no'>input</span>, <span class='kw'>quality</span> <span class='kw'>=</span> <span class='no'>input_q</span>, <span class='kw'>id</span> <span class='kw'>=</span> <span class='no'>input_names</span>)

<span class='co'># trim adapter</span>
<span class='no'>filtered</span> <span class='kw'>&lt;-</span> <span class='fu'>adapter_filter</span>(<span class='no'>my_read</span>, <span class='kw'>Rpattern</span> <span class='kw'>=</span> <span class='no'>adapter</span>, <span class='kw'>rc.R</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)

<span class='co'># watch the filtered sequences</span>
<span class='fu'><a href='http://www.rdocumentation.org/packages/ShortRead/topics/accessors'>sread</a></span>(<span class='no'>filtered</span>)</div><div class='output co'>#&gt;   A DNAStringSet instance of length 6
#&gt;     width seq
#&gt; [1]    43 TGGTCCGGTGTTCTGGCGGAATAGGTACAGTCCAGTAATTGCC
#&gt; [2]    43 TCCCGCAGACGCTGGGTCCGGAATGCCCTTTCTGAGCAGCTCC
#&gt; [3]    43 AGCCGTTTGACTTCGCGGAAAGTGAACTTAGATTCGGTCCTGA
#&gt; [4]    43 AACACGGTACTTCCACAGTCAACCCGCCGACTTGGAGAATTTA
#&gt; [5]    43 TTAGCCGGGCGGTTATTCCCCTAGTGATCTTACTAAGATTTGC
#&gt; [6]    43 AATACCTAAGCGAAGTGACAGATATGTTCGTCATTCATCCAGG</div><div class='input'>
<span class='co'># adapter in the second strand of paired-end reads is reverse-complemented,</span>
<span class='co'># with adapter in the end of sequence</span>
<span class='no'>adapterR</span> <span class='kw'>&lt;-</span> <span class='fu'>as.character</span>(<span class='fu'><a href='http://www.rdocumentation.org/packages/Biostrings/topics/reverseComplement'>reverseComplement</a></span>(<span class='fu'><a href='http://www.rdocumentation.org/packages/Biostrings/topics/DNAString-class'>DNAString</a></span>(<span class='st'>'ATCGACT'</span>)))
<span class='no'>adapterR</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='http://www.rdocumentation.org/packages/Biostrings/topics/DNAString-class'>DNAString</a></span>(<span class='no'>adapterR</span>)
<span class='no'>inputR</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='http://www.rdocumentation.org/packages/Biostrings/topics/reverseComplement'>reverseComplement</a></span>(<span class='no'>input</span>)


<span class='co'># create qualities of width 50</span>
<span class='no'>inputqR</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='random_qual.html'>random_qual</a></span>(<span class='fu'>c</span>(<span class='fl'>30</span>,<span class='fl'>40</span>), <span class='kw'>slength</span> <span class='kw'>=</span> <span class='fl'>6</span>, <span class='kw'>swidth</span> <span class='kw'>=</span> <span class='fl'>50</span>,
<span class='kw'>seed</span> <span class='kw'>=</span> <span class='fl'>10</span>, <span class='kw'>encod</span> <span class='kw'>=</span> <span class='st'>'Sanger'</span>)

<span class='no'>my_readR</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='http://www.rdocumentation.org/packages/ShortRead/topics/ShortReadQ-class'>ShortReadQ</a></span>(<span class='kw'>sread</span> <span class='kw'>=</span> <span class='no'>inputR</span>, <span class='kw'>quality</span> <span class='kw'>=</span> <span class='no'>inputqR</span>, <span class='kw'>id</span> <span class='kw'>=</span> <span class='no'>input_names</span>)

<span class='co'># trim adapter</span>
<span class='no'>filteredR</span> <span class='kw'>&lt;-</span> <span class='fu'>adapter_filter</span>(<span class='no'>my_readR</span>, <span class='kw'>Rpattern</span> <span class='kw'>=</span> <span class='no'>adapterR</span>)

<span class='co'># watch the filtered sequences</span>
<span class='fu'><a href='http://www.rdocumentation.org/packages/ShortRead/topics/accessors'>sread</a></span>(<span class='no'>filteredR</span>)</div><div class='output co'>#&gt;   A DNAStringSet instance of length 6
#&gt;     width seq
#&gt; [1]    50 ATCGACTGGCAATTACTGGACTGTACCTATTCCGCCAGAACACCGGACCA
#&gt; [2]    50 ATCGACTGGAGCTGCTCAGAAAGGGCATTCCGGACCCAGCGTCTGCGGGA
#&gt; [3]    50 ATCGACTTCAGGACCGAATCTAAGTTCACTTTCCGCGAAGTCAAACGGCT
#&gt; [4]    50 ATCGACTTAAATTCTCCAAGTCGGCGGGTTGACTGTGGAAGTACCGTGTT
#&gt; [5]    50 ATCGACTGCAAATCTTAGTAAGATCACTAGGGGAATAACCGCCCGGCTAA
#&gt; [6]    50 ATCGACTCCTGGATGAATGACGAACATATCTGTCACTTCGCTTAGGTATT</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Leandro Roser <a href='mailto:learoser@gmail.com'>learoser@gmail.com</a>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Leandro Roser, Fernán Agüero, Daniel Sánchez.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
