<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Remove full and partial adapters from a ShortReadQ object — adapter_filter • FastqCleaner</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Remove full and partial adapters from a ShortReadQ object — adapter_filter" />
<meta property="og:description" content="This program can remove adapters and partial 
adapters from 3' and 5', using the functions
trimLRPatterns 
 The program extends the methodology of
the  trimLRPatterns function of Biostrings,
being also capable of removing adapters present within reads and with other
additional otpions 
(e.g., threshold of minimum number of bases for trimming). 
For a given position in the read, the two Biostrings functions return TRUE 
when a match is present between a substring of the read and the adapter.
As trimLRPatterns , adapter_filter also selects
region and goes up to the end of the sequence in the corresponding flank 
as the best match. The default error rate is 0.2.
If several valid matches are found, the function removes the
largest subsequence. Adapters can be anchored or not.  
When indels are allowed, the second method uses the 'edit distance' between 
the subsequences and the adapter" />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">FastqCleaner</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.5.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Overview.html">An Introduction to FastqCleaner</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Remove full and partial adapters from a ShortReadQ object</h1>
    
    <div class="hidden name"><code>adapter_filter.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This program can remove adapters and partial 
adapters from 3' and 5', using the functions
<code><a href='https://rdrr.io/pkg/Biostrings/man/trimLRPatterns.html'>trimLRPatterns</a></code> 
 The program extends the methodology of
the  <code><a href='https://rdrr.io/pkg/Biostrings/man/trimLRPatterns.html'>trimLRPatterns</a></code> function of <span class="pkg">Biostrings</span>,
being also capable of removing adapters present within reads and with other
additional otpions 
(e.g., threshold of minimum number of bases for trimming). 
For a given position in the read, the two Biostrings functions return TRUE 
when a match is present between a substring of the read and the adapter.
As <code><a href='https://rdrr.io/pkg/Biostrings/man/trimLRPatterns.html'>trimLRPatterns</a></code> , adapter_filter also selects
region and goes up to the end of the sequence in the corresponding flank 
as the best match. The default error rate is 0.2.
If several valid matches are found, the function removes the
largest subsequence. Adapters can be anchored or not.  
When indels are allowed, the second method uses the 'edit distance' between 
the subsequences and the adapter</p>
    </div>

    <pre class="usage"><span class='fu'>adapter_filter</span>(<span class='no'>input</span>, <span class='kw'>Lpattern</span> <span class='kw'>=</span> <span class='st'>""</span>, <span class='kw'>Rpattern</span> <span class='kw'>=</span> <span class='st'>""</span>, <span class='kw'>rc.L</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>rc.R</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>first</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"R"</span>, <span class='st'>"L"</span>), <span class='kw'>with_indels</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>error_rate</span> <span class='kw'>=</span> <span class='fl'>0.2</span>, <span class='kw'>anchored</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>fixed</span> <span class='kw'>=</span> <span class='st'>"subject"</span>,
  <span class='kw'>remove_zero</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>checks</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>min_match_flank</span> <span class='kw'>=</span> <span class='fl'>3L</span>, <span class='no'>...</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>input</th>
      <td><p><code><a href='https://rdrr.io/pkg/ShortRead/man/ShortReadQ-class.html'>ShortReadQ</a></code> object</p></td>
    </tr>
    <tr>
      <th>Lpattern</th>
      <td><p>5' pattern (character or
<code><a href='https://rdrr.io/pkg/Biostrings/man/DNAString-class.html'>DNAString</a></code> object)</p></td>
    </tr>
    <tr>
      <th>Rpattern</th>
      <td><p>3' pattern (character or 
<code><a href='https://rdrr.io/pkg/Biostrings/man/DNAString-class.html'>DNAString</a></code> object)</p></td>
    </tr>
    <tr>
      <th>rc.L</th>
      <td><p>Reverse complement Lpattern? default FALSE</p></td>
    </tr>
    <tr>
      <th>rc.R</th>
      <td><p>Reverse complement Rpatter? default FALSE</p></td>
    </tr>
    <tr>
      <th>first</th>
      <td><p>trim first right('R') or left ('L') side of sequences
when both Lpattern and Rpattern are passed</p></td>
    </tr>
    <tr>
      <th>with_indels</th>
      <td><p>Allow indels? This feature is available only 
when the error_rate is not null</p></td>
    </tr>
    <tr>
      <th>error_rate</th>
      <td><p>Error rate (value in the range [0, 1] 
The error rate is the proportion of mismatches allowed between
the adapter and the aligned portion  of the subject.
For a given adapter A, the number of allowed mismatches between each 
subsequence s of A and the subject is computed as: error_rate * L_s,
where L_s is the length of the subsequence s</p></td>
    </tr>
    <tr>
      <th>anchored</th>
      <td><p>Adapter or partial adapter within sequence
(anchored = FALSE, default) or only in 3' and 5' terminals? 
(anchored = TRUE)</p></td>
    </tr>
    <tr>
      <th>fixed</th>
      <td><p>Parameter passed to 
<code><a href='https://rdrr.io/pkg/Biostrings/man/trimLRPatterns.html'>trimLRPatterns</a></code> 
Default 'subject',  ambiguities in the pattern only are interpreted 
as wildcard. See the argument fixed in 
<code><a href='https://rdrr.io/pkg/Biostrings/man/trimLRPatterns.html'>trimLRPatterns</a></code></p></td>
    </tr>
    <tr>
      <th>remove_zero</th>
      <td><p>Remove zero-length sequences? Default TRUE</p></td>
    </tr>
    <tr>
      <th>checks</th>
      <td><p>Perform checks? Default TRUE</p></td>
    </tr>
    <tr>
      <th>min_match_flank</th>
      <td><p>Do not trim in flanks of the subject,
if a match has min_match_flank of less length. Default 1L 
(only trim with &gt;=2 coincidences in a flank match)</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>additional parameters passed to
<code>trimLRpatterns</code></p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Edited <code><a href='https://rdrr.io/pkg/Biostrings/man/DNAString-class.html'>DNAString</a></code> or 
<code><a href='https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html'>DNAStringSet</a></code> object</p>
<p>Filtered <code><a href='https://rdrr.io/pkg/ShortRead/man/ShortReadQ-class.html'>ShortReadQ</a></code> 
object</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='fu'><a href='https://rdrr.io/r/base/library.html'>require</a></span>(<span class='st'>'Biostrings'</span>)</div><div class='output co'>#&gt; <span class='message'>Loading required package: Biostrings</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: BiocGenerics</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: parallel</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘BiocGenerics’</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:parallel’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,</span>
#&gt; <span class='message'>    clusterExport, clusterMap, parApply, parCapply, parLapply,</span>
#&gt; <span class='message'>    parLapplyLB, parRapply, parSapply, parSapplyLB</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:stats’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    IQR, mad, sd, var, xtabs</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:base’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    anyDuplicated, append, as.data.frame, basename, cbind, colnames,</span>
#&gt; <span class='message'>    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,</span>
#&gt; <span class='message'>    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,</span>
#&gt; <span class='message'>    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,</span>
#&gt; <span class='message'>    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,</span>
#&gt; <span class='message'>    union, unique, unsplit, which, which.max, which.min</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: S4Vectors</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: stats4</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘S4Vectors’</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:base’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    expand.grid</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: IRanges</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: XVector</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘Biostrings’</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:base’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    strsplit</span></div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/library.html'>require</a></span>(<span class='st'>'ShortRead'</span>)</div><div class='output co'>#&gt; <span class='message'>Loading required package: ShortRead</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: BiocParallel</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: Rsamtools</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: GenomeInfoDb</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: GenomicRanges</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: GenomicAlignments</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: SummarizedExperiment</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: Biobase</span></div><div class='output co'>#&gt; <span class='message'>Welcome to Bioconductor</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    Vignettes contain introductory material; view with</span>
#&gt; <span class='message'>    'browseVignettes()'. To cite Bioconductor, see</span>
#&gt; <span class='message'>    'citation("Biobase")', and for packages 'citation("pkgname")'.</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: DelayedArray</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: matrixStats</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘matrixStats’</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:Biobase’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    anyMissing, rowMedians</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘DelayedArray’</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:matrixStats’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:Biostrings’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    type</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:base’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    aperm, apply, rowsum</span></div><div class='input'>
<span class='co'># create 6 sequences of width 43</span>
<span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span>(<span class='fl'>10</span>)
<span class='no'>input</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='random_seq.html'>random_seq</a></span>(<span class='fl'>6</span>, <span class='fl'>43</span>)

<span class='co'># add adapter in 3' </span>
<span class='no'>adapter</span> <span class='kw'>&lt;-</span> <span class='st'>"ATCGACT"</span>

<span class='no'>input</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span>(<span class='no'>input</span>, <span class='fu'><a href='https://rdrr.io/r/base/character.html'>as.character</a></span>(<span class='fu'><a href='https://rdrr.io/pkg/Biostrings/man/DNAString-class.html'>DNAString</a></span>(<span class='no'>adapter</span>)))
<span class='no'>input</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html'>DNAStringSet</a></span>(<span class='no'>input</span>)

<span class='co'># create qualities of width 50</span>
<span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span>(<span class='fl'>10</span>)
<span class='no'>input_q</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='random_qual.html'>random_qual</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>30</span>,<span class='fl'>40</span>), <span class='kw'>slength</span> <span class='kw'>=</span> <span class='fl'>6</span>, <span class='kw'>swidth</span> <span class='kw'>=</span> <span class='fl'>50</span>,
<span class='kw'>encod</span> <span class='kw'>=</span> <span class='st'>'Sanger'</span>)

<span class='co'># create names</span>
<span class='no'>input_names</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='seq_names.html'>seq_names</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span>(<span class='no'>input</span>))

<span class='co'># create ShortReadQ object</span>
<span class='no'>my_read</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/ShortRead/man/ShortReadQ-class.html'>ShortReadQ</a></span>(<span class='kw'>sread</span> <span class='kw'>=</span> <span class='no'>input</span>, <span class='kw'>quality</span> <span class='kw'>=</span> <span class='no'>input_q</span>, <span class='kw'>id</span> <span class='kw'>=</span> <span class='no'>input_names</span>)

<span class='co'># trim adapter</span>
<span class='no'>filtered</span> <span class='kw'>&lt;-</span> <span class='fu'>adapter_filter</span>(<span class='no'>my_read</span>, <span class='kw'>Rpattern</span> <span class='kw'>=</span> <span class='no'>adapter</span>)

<span class='co'># look at the filtered sequences</span>
<span class='fu'><a href='https://rdrr.io/pkg/ShortRead/man/ShortRead-class.html'>sread</a></span>(<span class='no'>filtered</span>)</div><div class='output co'>#&gt;   A DNAStringSet instance of length 6
#&gt;     width seq
#&gt; [1]    43 TGGTCCGGTGTTCTGGCGGAATAGGTACAGTCCAGTAATTGCC
#&gt; [2]    43 TCCCGCAGACGCTGGGTCCGGAATGCCCTTTCTGAGCAGCTCC
#&gt; [3]    43 AGCCGTTTGACTTCGCGGAAAGTGAACTTAGATTCGGTCCTGA
#&gt; [4]    43 AACACGGTACTTCCACAGTCAACCCGCCGACTTGGAGAATTTA
#&gt; [5]    43 TTAGCCGGGCGGTTATTCCCCTAGTGATCTTACTAAGATTTGC
#&gt; [6]    43 AATACCTAAGCGAAGTGACAGATATGTTCGTCATTCATCCAGG</div><div class='input'>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Leandro Roser <a href='mailto:learoser@gmail.com'>learoser@gmail.com</a></p>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Leandro Roser, Fernán Agüero, Daniel Sánchez.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


