<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Remove left and right full and partial patterns — cutRseq • FastqCleaner</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Remove left and right full and partial patterns — cutRseq" />
<meta property="og:description" content="This set of programs are internal, 
and the function adapter_filter is recommended for trimming. 
The programs can remove adapters and partial 
adapters from 3' and 5'. The adapters can be anchored or not. 
When indels are allowed, the error rate consists in the edit distance. 
IUPAC simbols are allowed. The methods use the 
trimLRpatterns function
 of the Biostrings package, with some additions 
 to take into account e.g., partial adaptors.
 IUPAC symbols are allowed in all the cases. The present function 
also removes partial adapters, without the need of additional steps
 (for example, creating a padded adapter with 'Ns', etc). 
A similar result to the output of trimLRPatterns
can be obtained with the option anchored = TRUE.
When several matches are found, the function removes the subsequence 
that starts in the first match when cutRseq is used, or ends 
in the last match when cutLseq is used." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">FastqCleaner</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.5.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Overview.html">An Introduction to FastqCleaner</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Remove left and right full and partial patterns</h1>
    
    <div class="hidden name"><code>matching.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This set of programs are internal, 
and the function adapter_filter is recommended for trimming. 
The programs can remove adapters and partial 
adapters from 3' and 5'. The adapters can be anchored or not. 
When indels are allowed, the error rate consists in the edit distance. 
IUPAC simbols are allowed. The methods use the 
<code>trimLRpatterns</code> function
 of the <span class="pkg">Biostrings</span> package, with some additions 
 to take into account e.g., partial adaptors.
 IUPAC symbols are allowed in all the cases. The present function 
also removes partial adapters, without the need of additional steps
 (for example, creating a padded adapter with 'Ns', etc). 
A similar result to the output of <code><a href='https://rdrr.io/pkg/Biostrings/man/trimLRPatterns.html'>trimLRPatterns</a></code>
can be obtained with the option anchored = TRUE.
When several matches are found, the function removes the subsequence 
that starts in the first match when cutRseq is used, or ends 
in the last match when cutLseq is used.</p>
    </div>

    <pre class="usage"><span class='fu'>cutRseq</span>(<span class='no'>subject</span>, <span class='no'>Rpattern</span>, <span class='kw'>with.indels</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>fixed</span> <span class='kw'>=</span> <span class='st'>"subject"</span>,
  <span class='kw'>error_rate</span> <span class='kw'>=</span> <span class='fl'>0.2</span>, <span class='kw'>anchored</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>ranges</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>checks</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>min_match_flank</span> <span class='kw'>=</span> <span class='fl'>2L</span>, <span class='no'>...</span>)

<span class='fu'>cutLseq</span>(<span class='no'>subject</span>, <span class='no'>Lpattern</span>, <span class='kw'>with.indels</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>fixed</span> <span class='kw'>=</span> <span class='st'>"subject"</span>,
  <span class='kw'>error_rate</span> <span class='kw'>=</span> <span class='fl'>0.2</span>, <span class='kw'>anchored</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>ranges</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>min_match_flank</span> <span class='kw'>=</span> <span class='fl'>3L</span>, <span class='kw'>checks</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='no'>...</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>subject</th>
      <td><p><code><a href='https://rdrr.io/pkg/Biostrings/man/DNAString-class.html'>DNAString</a></code> or 
<code><a href='https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html'>DNAStringSet</a></code> object</p></td>
    </tr>
    <tr>
      <th>Rpattern</th>
      <td><p>3' pattern,
<code><a href='https://rdrr.io/pkg/Biostrings/man/DNAString-class.html'>DNAString</a></code> object</p></td>
    </tr>
    <tr>
      <th>with.indels</th>
      <td><p>Allow indels?</p></td>
    </tr>
    <tr>
      <th>fixed</th>
      <td><p>Parameter passed to 
<code>trimLRpatterns</code>
Default 'subject',  ambiguities in the pattern only are interpreted 
as wildcard. See the argument fixed in 
<code>trimLRpatterns</code></p></td>
    </tr>
    <tr>
      <th>error_rate</th>
      <td><p>Error rate (value in [0, 1]). 
The error rate is the proportion of mismatches allowed between
the adapter and the aligned portion of the subject.
For a given adapter A, the number of allowed mismatches between each 
subsequence s of A and the subject is computed as: error_rate * L_s,
where L_s is the length of the subsequence s.</p></td>
    </tr>
    <tr>
      <th>anchored</th>
      <td><p>Can the adapter or partial adapter be within 
the sequence? (anchored = FALSE)
or only in the terminal regions of the sequence? (anchored = TRUE).
Default TRUE (trim only flanking regions)</p></td>
    </tr>
    <tr>
      <th>ranges</th>
      <td><p>Return ranges? Default FALSE</p></td>
    </tr>
    <tr>
      <th>checks</th>
      <td><p>Perform internal checks? Default TRUE</p></td>
    </tr>
    <tr>
      <th>min_match_flank</th>
      <td><p>Do not trim in flanks of the subject,
if a match has min_match_flank of less length. Default 1L 
(only trim with &gt;=2 coincidences in a flank match)</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>additional parameters passed to
<code>trimLRpatterns]</code></p></td>
    </tr>
    <tr>
      <th>Lpattern</th>
      <td><p>5' pattern,
<code><a href='https://rdrr.io/pkg/Biostrings/man/DNAString-class.html'>DNAString</a></code> object</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Edited <code><a href='https://rdrr.io/pkg/Biostrings/man/DNAString-class.html'>DNAString</a></code> or 
<code><a href='https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html'>DNAStringSet</a></code> object</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/library.html'>library</a></span>(<span class='no'>Biostrings</span>)

<span class='no'>subject</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html'>DNAStringSet</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>'ATCATGCCATCATGAT'</span>,
<span class='st'>'CATGATATTA'</span>, <span class='st'>'TCATG'</span>, <span class='st'>'AAAAAA'</span>, <span class='st'>'AGGTCATG'</span>))

<span class='no'>Lpattern</span> <span class='kw'>&lt;-</span> <span class='no'>Rpattern</span> <span class='kw'>&lt;-</span> <span class='st'>'TCATG'</span>

<span class='fu'>cutLseq</span>(<span class='no'>subject</span>, <span class='no'>Lpattern</span>)</div><div class='output co'>#&gt; <span class='error'>Error in cutLseq(subject, Lpattern): could not find function "cutLseq"</span></div><div class='input'><span class='fu'>cutLseq</span>(<span class='no'>subject</span>, <span class='no'>Lpattern</span>, <span class='kw'>ranges</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</div><div class='output co'>#&gt; <span class='error'>Error in cutLseq(subject, Lpattern, ranges = TRUE): could not find function "cutLseq"</span></div><div class='input'><span class='fu'>cutRseq</span>(<span class='no'>subject</span>, <span class='no'>Rpattern</span>)</div><div class='output co'>#&gt; <span class='error'>Error in cutRseq(subject, Rpattern): could not find function "cutRseq"</span></div><div class='input'>

<span class='fu'>cutLseq</span>(<span class='no'>subject</span>, <span class='no'>Lpattern</span>, <span class='kw'>anchored</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</div><div class='output co'>#&gt; <span class='error'>Error in cutLseq(subject, Lpattern, anchored = FALSE): could not find function "cutLseq"</span></div><div class='input'><span class='fu'>cutLseq</span>(<span class='no'>subject</span>, <span class='no'>Lpattern</span>,  <span class='kw'>error_rate</span> <span class='kw'>=</span> <span class='fl'>0.2</span>)</div><div class='output co'>#&gt; <span class='error'>Error in cutLseq(subject, Lpattern, error_rate = 0.2): could not find function "cutLseq"</span></div><div class='input'><span class='fu'>cutLseq</span>(<span class='no'>subject</span>, <span class='no'>Lpattern</span>,  <span class='kw'>error_rate</span> <span class='kw'>=</span> <span class='fl'>0.2</span>,
<span class='kw'>with.indels</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</div><div class='output co'>#&gt; <span class='error'>Error in cutLseq(subject, Lpattern, error_rate = 0.2, with.indels = TRUE): could not find function "cutLseq"</span></div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Leandro Roser <a href='mailto:learoser@gmail.com'>learoser@gmail.com</a></p>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Leandro Roser, Fernán Agüero, Daniel Sánchez.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


